### ISSUE TITLE
feat: Using authz-keycloak plugin, I want to add claims to an authz request

### ISSUE THREAD

**Issue by Java-Superman** [apache/apisix#11094]
### Add claim_token to authz request

Using the [JS Policy](https://www.keycloak.org/docs/24.0.1/authorization_services/#_policy_js) in Keycloak, I need to know the parameters in the URL to decide if the user has permission. An example would be how to handle the path `/region/1234/customers`, where the user must belong to `group.regionId=1234`

Claims can be added to the authz request like this [example.](https://www.keycloak.org/docs/24.0.1/authorization_services/#_service_pushing_claims)
```
curl -X POST \
  http://${host}:${port}/realms/${realm}/protocol/openid-connect/token \
  --data "grant_type=urn:ietf:params:oauth:grant-type:uma-ticket" \
  --data "claim_token=ewogICAib3JnYW5pemF0aW9uIjogWyJhY21lIl0KfQ==" \
  --data "claim_token_format=urn:ietf:params:oauth:token-type:jwt" \
  --data "client_id={resource_server_client_id}" \
  --data "client_secret={resource_server_client_secret}" \
  --data "audience={resource_server_client_id}"
```

It would be nice if the plugin added the uri (possibly other things) to the claim_token.

`claim_token = base64( { "uri" : $request_uri } )`  put in [authz-keycloak.lua](https://github.com/apache/apisix/blob/master/apisix/plugins/authz-keycloak.lua#L626) 

In **Java** using Keycloak's ServletPolicyEnforcerFilter, an example configuration file for claims looks like [this](https://www.keycloak.org/docs/latest/authorization_services/index.html#obtaining-information-from-the-http-request).
```
{
  ....
  "paths": [
    {
      "path": "/protected/resource",
      "claim-information-point": {
        "claims": {
          "claim-from-request-parameter": "{request.parameter['a']}",
          "claim-from-header": "{request.header['b']}",
          "claim-from-cookie": "{request.cookie['c']}",
          "claim-from-remoteAddr": "{request.remoteAddr}",
  ...
}
```

### Use the request path for the permission

For **authz-keycloak.lua**, I believe the _permission_ config parameter could be optional if the authz request used `permission=$url` and  `permission_resource_matching_uri=true` when _permission_ is not supplied (see [here](https://www.keycloak.org/docs/latest/authorization_services/index.html#_service_obtaining_permissions)).  This would allow permissions to be controlled based on the path within keycloak and not have the additional step of adding all the permissions to APISIX.

NOTE: Relates to https://github.com/apache/apisix/issues/7190

**Comment by shreemaan-abhishek** [comment]
how would these claims be passed to apisix? I mean how would apisix know which claims to pass?

**Comment by shreemaan-abhishek** [comment]
> Defining it within the attributes of the [authz-keycloak](https://apisix.apache.org/docs/apisix/plugins/authz-keycloak/) plugin. I'm not a lua dev, but something like:
> 
> `claims : { "uri" : "var.uri", "myparam" : "var.arg_myparam" } `
> 
> where they would be evaluated at runtime. I'll have to defer to you guys on how the properties above could be evaluated at runtime (finite set of things/eval method).

hardcoding them doesn't make sense to me.

**Comment by Java-Superman** [comment]
@shreemaan-abhishek 
> hardcoding them doesn't make sense to me.

I agree, but I'm suggesting the attribute values within the claims object would need to be **evaluated** within the lua script during a request.

For the example: 
```
"plugins": {
  "authz-keycloak": {
    "claims": {
      "uri": "var.uri",
      "myparam": "var.arg_myparam"
    }
  }
}
```
During an http request

- var.uri - evaluates to the current uri (e.g. /foo?myparam=1234)
- var.arg_myparam - evaluates to the current query string myparam (e.g. 1234)

The claim_token passed in the lua script would be:

`claim_token = base64( stringify( { "uri": "/foo?myparam=1234", "myparam": "1234" } ) )`

I'm not sure of the best way the attribute values could/should be evaluated.  One possibility is to parse the string (e.g. "var.uri") and include the result in the claim.

**Comment by shreemaan-abhishek** [comment]
ah okay, makes sense to me now. I'm a noob when it comes to oidc/keycloak stuff.

This looks like a good to have but a big feature nevertheless. Let's wait for the community to drop their feedback.
